022-feature-gates.md

```diff
@@ -0,0 +1,125 @@
+# Feature Gates
+
+This proposal suggests to add support for Feature Gates in the Strimzi operators.
+Feature Gates should give us additional options how to control and mature different behaviors in the operators.
+
+## Current situation
+
+Currently, any features or behaviors of the operator are controlled by one of these two mechanisms:
+
+* The Strimzi custom resources (as a form of an API)
+* The state of the Kubernetes resources
+
+The following examples show how are these mechanisms used to give a better idea:
+
+* The `strimzi.io/use-connector-resources` annotation in the `KafkaConnect` custom resource defines whether the Connector Operator should be enabled or not for given Kafka Connect resource.
+* The `useServiceDnsDomain` in the listener configuration of the `Kafka` custom resource defines what kind of address should be used for given listener.
+  Whether fully qualified DNS name or not.
+* When the operator during Kafka upgrade sees that the old pods are still using the TLS sidecars, it will automatically loosen the TLS configuration to make sure the upgrade can be completed.
+
+## Motivation
+
+In many cases, the mechanisms described above work fine and are well suited.
+But there are also limitations:
+
+* Changes to the custom resources APIs have permanent character.
+  We can remove fields from the CRDs only when changing the API versions.
+  So adding fields to the Kafka CR is not a good option to gradually roll out changes or change a default behavior over multiple releases.
+* Sometimes, you also want to just change the behavior of the operator in general.
+  But you want to avoid having endless list of `enableFeatureX`, `disableFeatureY`, or `useZ` fields in the custom resources.
+* The use of the state of the existing Kubernetes resources works well for one-of operations.
+  That fits well for upgrades or for triggering actions such as restarting a pod.
+  But not for features which change the operator behavior during the whole run of the operator.
+
+This shows the there might be space for other ways how to enable or disable features.
+
+## Proposal
+
+Strimzi operators should support _feature gates_ to gate selected functionality and switch it on or off.
+Each feature gate will have a default state (enabled or disabled by default) and users will be able to manually enable or disable them.
```

Comment:
Not to over-engineer here, but is there ever a case where we may want enums instead of booleans?
I think Booleans can cover almost all use cases, but I suppose there might be interesting scenarios where whole different options could be considered?

Reply:
I would probably prefer to avoid this. I think that it would make it more complicated not just to use them in the code but also make the user experience worse and more complicated.

Reply:
Makes sense